[Sema] FixIt for `override func` that should be `override var` and vice-versa

Rajveer100 opened the pull request:
Fixes #57499 

This adds `override_property_not_method` and `override_method_not_property` notes by performing
a member lookup for the respective diagnostics.

Rajveer100 commented:
@AnthonyLatsis I have managed to get the `diagnosis` to work, and getting the desire output with the note...Please review the changes and correct any other things required apart from adding `tests`.

Rajveer100 commented:
How to `debug a Fix-It`, meaning how to show the fixit and see the new form?
(Like we would normally click it and see the code change in the IDE)

AnthonyLatsis commented:
> How to debug a fix-it

See https://github.com/apple/swift/blob/main/docs/Diagnostics.md#diagnostic-verifier

Rajveer100 commented:
Alright, I managed to add the fixit as well, is there anything else I must add to this:

```
  override func canBecomeFirstResponder() -> Bool {}
           ~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           var canBecomeFirstResponder: Bool
```

xedin commented on lib/Sema/TypeCheckDeclPrimary.cpp:
```diff
@@ -2147,6 +2147,31 @@ class DeclChecker : public DeclVisitor<DeclChecker> {
                 .highlight(OA->getLocation());
+            //If the override is a property-decl, check if there exists
+            //a no-args func-decl in the base class, if yes, provide a fix-it.
+            auto declName = VD->getBaseName();
```
A couple of this here:

- There is no need to use constraint system for this. You need to extract nominal type (i.e. struct/enum/class) where it's declared via `DC->getSelfNominalTypeDecl()` - using this declaration it possible to lookup members via `lookupDirect` method.
-  The information about number of parameters and result type could be extracted from `FuncDecl`

xedin commented on lib/Sema/TypeCheckDeclPrimary.cpp:
```diff
@@ -3177,6 +3202,25 @@ class DeclChecker : public DeclVisitor<DeclChecker> {
+              if (!inherited.empty()) {
+                auto superClass = inherited.front().getType();
+                constraints::ConstraintSystem CS(DC, None);
+                auto lookupResult = CS.performMemberLookup(swift::constraints::ConstraintKind::ValueMember, DeclNameRef(declName), superClass, FunctionRefKind::SingleApply, CS.getConstraintLocator({}), false);
```
Same comment as about - the lookup should be done without constraint system involvement.

xedin reviewed (changes requested):
I have left some comments inline and please add test-cases to `test/decl/class/override.swift` and `test/decl/inherit/override.swift` that check that newly added fix-its and notes.

Rajveer100 commented:
Hey, thanks! @xedin 
I will make the changes and add the test files.

xedin commented:
Happy to help! Please press "re-request view" once you feel like that PR is ready for another round of review!

Rajveer100 commented:
@xedin What would be the right way to add tests to the respective files you mentioned?

xedin commented:
All new tests are usually appended to the end of the test file, you should mention Github issue and/or radar number and wrap in either do or func declarations to avoid polluting global namespace with names that are local to tests.

Rajveer100 commented:
Alright!

Rajveer100 commented:
@AnthonyLatsis @xedin 

The `-verify` frontend flag which is told to be used in order to check the expected `note` or `fixit` (in this case) must be added or passed exactly where in the Xcode workspace?

AnthonyLatsis commented:
The commands that are run for a given test file are usually specified at the top of the file using `RUN:` directives. The files Pavel mentioned should already be passing `verify` in at least one of those invocations. Also, you should not be attempting to run tests in Xcode.

Rajveer100 commented:
I am aware of running tests, as I had done them previously in my pull requests via this command:

```
utils/run-test --lit ../llvm-project/llvm/utils/lit/lit.py \
  ../build/Ninja-RelWithDebInfoAssert/swift-macosx-$(uname -m)/test-macosx-$(uname -m) \
  --filter="MyTest"
```
But regarding the flags, I thought some extra work was needed as this is the first time I am adding test cases, my doubt was whether I should use certain `-flag` in that command in terminal, or is it some flag that needs to be added in my workspace and that would `throw` something and I don't know what exactly the `RUN` command does as I had seen it before too.

So what would be the appropriate command to execute, let's say, for the file `test/decl/class/override.swift`:

Would it be the same way as the command I used earlier (mentioned on top of this comment)?

AnthonyLatsis commented:
If the `-verify` flag is present in one of the `RUN:` lines (which it ought to in these particular files), then no extra work is needed. `lit` does not do anything special — it simply executes the `RUN:` lines. Each test file is responsible for supplying its own testing commands using this directive.

Rajveer100 commented:
Ahh I see! Thanks!

Rajveer100 commented on lib/Sema/TypeCheckDeclPrimary.cpp:
```diff
@@ -3213,9 +3213,9 @@ class DeclChecker : public DeclVisitor<DeclChecker> {
                   auto candidateType = candidate->getInterfaceType()->getRValueType();
                   if (candidateType->isEqual(declReturnType))
                     FD->diagnose(diag::override_method_in_lieu_of_property, declName.getIdentifier())
-                        .fixItReplace(FD->getSourceRange(),
```
```
error: method does not override any method from its superclass
  override func canBecomeFirstResponder() -> Bool {
  ~~~~~~~~      ^
note: did you mean to override the property 'canBecomeFirstResponder'?
  override func canBecomeFirstResponder() -> Bool {
  ~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  override var canBecomeFirstResponder: Bool { <#code#> }
```

Rajveer100 replied:
This works as expected.

Rajveer100 commented on lib/Sema/TypeCheckDeclPrimary.cpp:
```diff
@@ -3213,9 +3213,9 @@ class DeclChecker : public DeclVisitor<DeclChecker> {
-                                      "var " + declName.getIdentifier().str().str() + ": " + declReturnType.getString() + " { <#code#> }")
-                        .fixItRemove(FD->getBodySourceRange());
+                        .fixItRemove(FD->getBodySourceRange())
+                        .fixItReplace(FD->getSourceRangeIncludingAttrs(),
```
```
error: property does not override any property from its superclass
  override var canBecomeFirstResponder: Int {
  ~~~~~~~~     ^
note: did you mean to override the method 'canBecomeFirstResponder'?
  override var canBecomeFirstResponder: Int {
               ^~~~~~~~~~~~~~~~~~~~~~~
               override func canBecomeFirstResponder() -> Int { <#code#> }
```

Rajveer100 replied:
In this case, the replacement starts after `override var`...

xedin replied:
I left a comment about this in previous fix-it location.

Rajveer100 commented:
@xedin Could you please review this, I have mentioned few in line comments as well to describe certain issues.

xedin commented on lib/Sema/TypeCheckDeclPrimary.cpp:
```diff
@@ -2147,6 +2147,27 @@ class DeclChecker : public DeclVisitor<DeclChecker> {
+            //If there is an inherited context, perform member lookup.
+            if (!inherited.empty()) {
+              auto superClass = inherited.front().getType();
+              auto lookupResult = superClass->getAnyNominal()->lookupDirect(declName);
```
Ok, I realized that I did mislead you a bit - you actually want to use `TypeChecker::lookupMember` with `DC->getSelfClassDecl()` because the member could be located farther in the superclass chain too.

Rajveer100 replied:
That's fine, that led me to figuring out more things and increased my knowledge a ton! 

Although, could you give me a brief explanation for each type of lookup, i.e., the `constraints`, `lookupdirect` and `lookupmember`, according to my understanding, the constraint system performs a more generalised form of lookup and is too overkill for this context. Other two, I am not too sure about...

xedin replied:
`lookupDirect` finds only members directly on the given type and its extensions, `lookupMember` is going to look at all the supertypes too, there is also unqualified lookup (`TypeChecker::lookupUnqualified`) it would lookup declarations with the given name in the given declaration context.

Rajveer100 replied:
Ahh! Thanks for the info! I will make the changes!

Rajveer100 replied:
@xedin Correct me if I am wrong, I have tried the following ways to fetch the required member, but none of them give any results: 

`TypeChecker::lookupMember(DC->getSelfClassDecl(), declReturnType, DeclNameRef(declName))`
`TypeChecker::lookupMember(DC, declReturnType, DeclNameRef(declName))`

AnthonyLatsis replied:
You may realize the mistake if you read the documentation for `TypeChecker::lookupMember`.

Rajveer100 replied:
Thanks @AnthonyLatsis, will look into the same!

Rajveer100 replied:
@AnthonyLatsis 
I have tried figuring out with several different possibilities and also looked through use cases and examples across the codebase, but didn't quite figure out yet. All of them throw the null pointer error or similar.

AnthonyLatsis replied:
Need help?

Rajveer100 replied:
Well, not yet! I wanna try my best!

Rajveer100 replied:
Regarding the `TypeChecker::lookupMember`, how does it exactly lookup for the member, in the sense that, the first parameter I suppose would take the class decl, i.e., `DC->getSelfClassDecl()`, the second parameter, which takes the `Type`, would be the exact return type, as I tried above or would it be of some other form due to which it fails to fetch any results (first and third parameter I suppose I have passed correctly)?

xedin replied:
Sometimes it's helpful to take a look at some existing call sites i.e. TypeCheckAttr.cpp has a lot of uses. First argument is a declaration context that is requesting the member, `DC` in your case, the second argument is the type of class you are trying to lookup members for, which is `DC->getSelfClassDecl()->getDeclaredType()`.

AnthonyLatsis replied:
https://github.com/apple/swift/blob/ad3fec4e7e73be7ad077fb66e9ce3edbc2fe7240/lib/Sema/TypeChecker.h#L886-L896

Rajveer100 replied:
Thanks for helping out! @xedin @AnthonyLatsis!

Rajveer100 replied:
> https://github.com/apple/swift/blob/ad3fec4e7e73be7ad077fb66e9ce3edbc2fe7240/lib/Sema/TypeChecker.h#L886-L896
> 
> 
> 
> `type` is not the return type of the members it should look for.

I had realised that pretty early, cause initially, yes, I was passing the function return type, but was stuck after that, thanks!

Rajveer100 replied:
`Assertion failed: (Val && "isa<> used on a null pointer")`

Rajveer100 replied:
`Assertion failed: (Val && "isa<> used on a null pointer"), function doit, file Casting.h, line 104.`

Still the same error.

Rajveer100 replied:
Seems to be a problem with this line, `genericParams = cast<ExtensionDecl>(this)->getGenericParams();`

xedin replied:
Looks like you'd actually have to get this under the debugger to see what is going on, I don't know why lookup would be crashing like this.

Rajveer100 replied:
Hmm.

Rajveer100 replied:
@xedin 
Hey! I managed to debug it!

So it is a function parameter-cased issue. Essentially, when there aren't any parameters in the function, it crashes!

Rajveer100 replied:
Ignore the last deleted comment, I hadn't changed the if else part.

Rajveer100 replied:
Meanwhile, debugging to what could be the cause.

xedin replied:
FYI that `DC->getSelfClassDecl()->getSuperclass()` could produce an `Type()` which is effectively a `nullptr`.

Rajveer100 replied:
Well, but, when I did `llvm::errs() << DC->getSelfClassDecl()->getSuperclass();`, it does show class 'C' which in my case is the super class.

xedin replied:
That's fine, I'm just saying that it could be.

xedin commented on lib/Sema/TypeCheckDeclPrimary.cpp:
```diff
@@ -2147,6 +2147,27 @@ class DeclChecker : public DeclVisitor<DeclChecker> {
+              for (auto& candidate : lookupResult) {
+                //If the parameter list is empty, compare the return types
+                //of the candidates with the declaration type.
+                if (getParameterList(candidate)->size() == 0) {
```
I think a safer option here would be to determine whether there is only ne member with no parameters and a matching result type.

Rajveer100 replied:
You mean to say, we should have an additional check before parameter size, that the lookupResult should have just one candidate? (i.e. `lookupResult.size() == 0` )?

xedin replied:
You need to determine that there is a single declaration that the new member is trying to override for that you need to make sure that there are no parameters and the result type matches exactly, you can use `llvm::find_first` for that walking over the results.

Rajveer100 replied:
Hmm, I will see that!

xedin commented on lib/Sema/TypeCheckDeclPrimary.cpp:
```diff
@@ -2147,6 +2147,27 @@ class DeclChecker : public DeclVisitor<DeclChecker> {
+                  auto candidateReturnType = candidate->getResultTypeRepr();
+                  if (candidateReturnType->getKind() == declType->getKind())
+                    VD->diagnose(diag::override_property_in_lieu_of_method, declName.getIdentifier())
+                        .fixItReplace(VD->Decl::getSourceRangeIncludingAttrs(), "override func " + declName.getIdentifier().str().str() + "() -> " + VD->getInterfaceType()->getRValueType().getString() + " { <#code#> }");
```
You cannot replace the whole declaration, what you need to do is to find location of `var` and replace that then get location of where the type starts (after `:`) and replace that with your new function type.

xedin commented on test/decl/class/override.swift:
```diff
@@ -414,3 +414,28 @@ open class OpenDerivedFinal : OpenBase {
+
+  class Inherited: BaseClass {
+    override var someFuncX: Int { // expected-error {{property does not override any property from its superclass}}
+    // expected-note@-1 {{did you mean to override the method 'someFuncX'?}}
```
Fix-its have to be validated as well via {{<loc>-<loc>=<new code>}} for example - https://github.com/apple/swift/blob/main/test/Constraints/diagnostics.swift#L98

Rajveer100 replied:
Yeah, there were some things that needed to be fixed as you suggested regarding fix-its, that's why I left it for the end.

Rajveer100 commented:
@xedin 
Request you to look into this [thread](https://github.com/apple/swift/pull/63858#discussion_r1125411609).

Rajveer100 commented:
@xedin 
Been quite long since we last spoke about this issue! Any advice based on the last conversation we had, if I correctly remember I had used the overkill `constraint system` for the `lookup`, which was still working correctly, but as you suggested later to use `lookup direct` which is where the last problem occurred as [here](https://github.com/apple/swift/pull/63858#discussion_r1123808782).

xedin commented:
I don’t have any other suggestions besides running your changes under debugger and trying to figure out the root cause of the crash, the approach we have discussed should be sound, it’s the a matter of figuring out how exactly to use the APIs.

AnthonyLatsis commented:
@Rajveer100 Are you still working on this?

Rajveer100 commented:
Does this issue still exist, if you have already reproduced in the latest Swift version?

It's been long since I actively worked on it, but if someone has a PR I don't have a problem :)

Rajveer100 committed:
[Sema] FixIt for `override func` that should be `override var` and vice-versa

Resolves #57499

This adds `override_property_not_method` and `override_method_not_property` notes by performing
a member lookup for the respective diagnostics.

Rajveer100 commented:
@xedin 
Could you review this as well :)

AnthonyLatsis reviewed (commented):
I think a better approach is to handle this case in `OverrideMatcher` (where name lookup is already done for us) during `OverrideCheckingAttempt::BaseName`, and to ignore any other differences between the two declarations. Just always emit a tailored diagnostic for var vs. func, and leave the fix-its for another time. Quality fix-its are a deeper rabbit hole as things stand because the current logic is both exclusive and dependent on declaration kind equality, and there are plenty of rules aside from type matching for whether one property or method is a valid override of another.

___

That said, a great improvement to your current change would be to extract the diagnosis into a function.
